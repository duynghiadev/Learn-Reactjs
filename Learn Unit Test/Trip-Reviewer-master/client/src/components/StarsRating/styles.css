.stars-rating-container .star-empty {
  color: black;
}

.stars-rating-container .error-empty,
.stars-rating-container .error-label {
  color: red;
}

.stars-rating-container .star-empty:before {
  font-size: 39px;
}

.stars-rating-container .star-full {
  color: gold;
  margin-top: 6px;
}

.stars-rating-container .star-full:before {
  font-size: 27px;
}

.stars-rating-container .rating-label {
  display: block;
  font-size: 24px;
  margin-bottom: 5px;
}
