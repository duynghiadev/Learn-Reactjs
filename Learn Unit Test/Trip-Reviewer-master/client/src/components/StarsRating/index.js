import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Rating from 'react-rating';

import './styles.css';

class StarsRating extends PureComponent {
  static getLabel(rating) {
    switch (rating) {
      case 1:
        return 'Hated it!';
      case 2:
        return 'Disliked it!';
      case 3:
        return 'It was ok.';
      case 4:
        return 'Liked it!';
      case 5:
        return 'Loved it!';
      default:
        return 'Rate your trip:';
    }
  }

  state = {
    hower: null,
  };

  handleRating = rating => {
    this.props.onChange(rating);
  };

  handleHower = hower => {
    this.setState({ hower });
  };

  render() {
    const { hower } = this.state;
    const { rating, error } = this.props;

    const emptySymbol = (
      <i className={`far fa-star star-empty${error && !hower ? ' error-empty' : ''}`} />
    );
    const fullSymbol = (
      <i className={`fas fa-star star-full${error && !hower ? ' error-full' : ''}`} />
    );

    return (
      <div className="stars-rating-container">
        <label className={`rating-label${error && !hower ? ' error-label' : ''}`}>
          {StarsRating.getLabel(hower || rating)}
        </label>
        <Rating
          initialRating={rating}
          emptySymbol={emptySymbol}
          fullSymbol={fullSymbol}
          onChange={this.handleRating}
          onHover={this.handleHower}
        />
      </div>
    );
  }
}

StarsRating.propTypes = {
  rating: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
};

StarsRating.defaultProps = {
  error: false,
};

export default StarsRating;
