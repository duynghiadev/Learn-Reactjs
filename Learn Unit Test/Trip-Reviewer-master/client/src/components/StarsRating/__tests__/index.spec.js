import React from 'react';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';

import StarsRating from '..';

jest.mock('react-rating', () => 'Rating');

const mockComponent = props => <StarsRating {...props} />;

describe('<StarsRating />', () => {
  it('should render component', () => {
    const tree = create(mockComponent()).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with error', () => {
    const props = { error: true };
    const tree = create(mockComponent(props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should change state correctly when handleHower is called', () => {
    const instance = mount(mockComponent()).instance();
    const hower = 3;

    instance.handleHower(hower);

    expect(instance.state.hower).toEqual(hower);
  });

  it('should call onChange when handleRating is called', () => {
    const onChange = jest.fn();
    const props = { onChange };
    const rating = 3;

    const instance = mount(mockComponent(props)).instance();

    instance.handleRating(rating);

    expect(onChange).toHaveBeenCalledWith(rating);
  });

  it('should return correct value when getLabel is called', () => {
    const one = StarsRating.getLabel(1);
    const two = StarsRating.getLabel(2);
    const three = StarsRating.getLabel(3);
    const four = StarsRating.getLabel(4);
    const five = StarsRating.getLabel(5);
    const others = StarsRating.getLabel(100);

    expect(one).toEqual('Hated it!');
    expect(two).toEqual('Disliked it!');
    expect(three).toEqual('It was ok.');
    expect(four).toEqual('Liked it!');
    expect(five).toEqual('Loved it!');
    expect(others).toEqual('Rate your trip:');
  });
});
