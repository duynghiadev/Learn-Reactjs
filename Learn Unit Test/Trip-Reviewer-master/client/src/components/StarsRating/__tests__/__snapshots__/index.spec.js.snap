// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarsRating /> should render component 1`] = `
<div
  className="stars-rating-container"
>
  <label
    className="rating-label"
  >
    Rate your trip:
  </label>
  <Rating
    emptySymbol={
      <i
        className="far fa-star star-empty"
      />
    }
    fullSymbol={
      <i
        className="fas fa-star star-full"
      />
    }
    initialRating={undefined}
    onChange={[Function]}
    onHover={[Function]}
  />
</div>
`;

exports[`<StarsRating /> should render component with error 1`] = `
<div
  className="stars-rating-container"
>
  <label
    className="rating-label error-label"
  >
    Rate your trip:
  </label>
  <Rating
    emptySymbol={
      <i
        className="far fa-star star-empty error-empty"
      />
    }
    fullSymbol={
      <i
        className="fas fa-star star-full error-full"
      />
    }
    initialRating={undefined}
    onChange={[Function]}
    onHover={[Function]}
  />
</div>
`;
